File: python_implementation/checksum.py

def fletcher(data = [], k=16):

    sum1 = 0
    sum2 = 0
    mod = 255

    if k == 16:
        mod = 255
    elif k == 32:
        mod = 65535
    elif k == 64:
        mod = 4294967295

    for i in range(len(data)):
        sum1 = (sum1 + data[i]) % mod
        sum2 = (sum2 + sum1) % mod

    checksum = (sum2 << 8) | sum1
    return bin(checksum)[2:].zfill(k)


def binaryStringToByteArray(binaryString):
    byteArray = []
    for i in range(0, len(binaryString), 8):
        byteArray.append(int(binaryString[i:i+8], 2))
    return byteArray

def decodeFletcher(content, k=16):
    checksum = content[-k:]
    frame = content[:-k]
    data = binaryStringToByteArray(frame)
    expectedChecksum = fletcher(data, k)
    return expectedChecksum == checksum



checksum = fletcher(binaryStringToByteArray('0101001'), 32)
answer = "Si" if decodeFletcher('0101001' + checksum, 16) else "No"

print(f"Fletcher-16 Checksum: {checksum}")
print(f"Codigo valido {answer}")

File: typescript_implementation/checksum.ts

const fletcher = (data: number[], k = 16) => {
	
  let sum1 = 0;
	let sum2 = 0;
	
	let mod = 255;
	// Configurando el algoritmo dependiendo de K
	if (k == 16) mod = 255
	else if (k == 32) mod = 65535
	else if (k == 64) mod = 4294967295

  for (let i = 0; i < data.length; i++) {
    sum1 = (sum1 + data[i]) % mod;
		sum2 = (sum2 + sum1) % mod;
  }
	const checksum = (sum2 << 8) | sum1;
	let checksumString = checksum.toString(2).toUpperCase();
	// Agregando ceros a la izquierda
	while ((checksumString.length % k) != 0) {
		checksumString = '0' + checksumString;
	}
  return checksumString;
}

const binaryStringToByteArray = (str: string) => {
	const byteArray = [];
	for (let i = 0; i < str.length; i += 8) {
		byteArray.push(parseInt(str.slice(i, i + 8), 2));
	}
	return byteArray;
}


const decodeFletcher = (content: string, k = 16) => {
	// checksum is the last k bits
	const checksum = content.slice(content.length - k);
	const frame = content.slice(0, content.length - k);
	const data = binaryStringToByteArray(frame);
	const expectedChecksum = fletcher(data, k);
	return expectedChecksum === checksum;
}

// Example usage:
const checksum = fletcher(binaryStringToByteArray('0101001'), 16);
console.log("Fletcher-16 Checksum:", checksum);
console.log("Codigo valido: ", decodeFletcher('0101001' + checksum, 16) ? "Si" : "No");

